File: engine/src/main/java/com/terralite/engine/terrain/BlockState.java
package com.terralite.engine.terrain;

import com.terralite.core.registry.ResourceId;

import java.util.Map;
import java.util.Objects;

public record BlockState(ResourceId id, Map<String, String> properties) {
    public static final BlockState AIR = new BlockState(ResourceId.id("terralite:air"));

    public BlockState(ResourceId id) {
        this(id, Map.of());
    }

    public BlockState {
        Objects.requireNonNull(id, "id");
        properties = Map.copyOf(Objects.requireNonNull(properties, "properties"));
    }

    public static BlockState of(String id) {
        return new BlockState(ResourceId.id(id));
    }

    public BlockState with(String property, String value) {
        Objects.requireNonNull(property, "property");
        Objects.requireNonNull(value, "value");
        if (property.isBlank()) {
            throw new IllegalArgumentException("Block state property cannot be blank");
        }
        if (value.isBlank()) {
            throw new IllegalArgumentException("Block state value cannot be blank");
        }

        java.util.LinkedHashMap<String, String> next = new java.util.LinkedHashMap<>(properties);
        next.put(property, value);
        return new BlockState(id, next);
    }

    public String property(String property) {
        return properties.get(Objects.requireNonNull(property, "property"));
    }

    public boolean isAir() {
        return this.equals(AIR);
    }
}

File: engine/src/commonMain/kotlin/com/terralite/engine/terrain/BlockState.kt
package com.terralite.engine.terrain

import com.terralite.core.registry.ResourceId

@JvmRecord
data class BlockState(val id: ResourceId, val properties: Map<String, String>) {
    constructor(id: ResourceId) : this(id, emptyMap())

    init {
        @Suppress("SENSELESS_COMPARISON")
        require(id != null) { "id" }
    }

    companion object {
        @JvmField val AIR: BlockState = BlockState(ResourceId.id("terralite:air"))

        @JvmStatic fun of(id: String): BlockState = BlockState(ResourceId.id(id))
    }

    fun with(property: String, value: String): BlockState {
        require(property.isNotBlank()) { "Block state property cannot be blank" }
        require(value.isNotBlank()) { "Block state value cannot be blank" }
        val next = LinkedHashMap(properties)
        next[property] = value
        return BlockState(id, next)
    }

    fun property(property: String): String? = properties[property]

    fun isAir(): Boolean = this == AIR
}
